### PR TITLE
tests: add a new stress test for mutex

### PR DESCRIPTION
**Pull request by chen-png**
Add a new stress test to validate mutex lock, unlock and semaphore take, give
operations. In this test, there are two threads, they will compete 
to get a common mutex and semaphore, and it will run about 1 hour
except for meeting an error.

this is mainly for #39575, for some real applications, it needs
long time to find out those potential issues.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

### CODE REVIEW

**Comment by chen-png** [comment]
hi @andyross I started a timer like this "k_timer_start(&timer, K_HOURS(1), K_NO_WAIT);", then it should be expired after 1 hour, but **why it only took several minutes to be expired when I tested it on qemu_x86**? when I tested it on up_squared, it worked well.

**Comment by jhedberg** [comment]
>  I started a timer like this "k_timer_start(&timer, K_HOURS(1), K_NO_WAIT);", then it should be expired after 1 hour, but why it only took several minutes to be expired when I tested it on qemu_x86? when I tested it on up_squared, it worked well.

If I were to guess, that's because of integer overflow e.g. due to `CONFIG_TIMEOUT_64BIT` not being set.  You should probably try building for `qemu_x86_64` instead of `qemu_x86`.

File: tests/stress_tests/mutex_stress/src/main.c [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+K_THREAD_STACK_ARRAY_DEFINE(thread_stack, THREAD_COUNT, THREAD_STACK_SIZE);
+
+struct k_thread sem_thread;
+K_THREAD_STACK_DEFINE(sem_thread_stack, THREAD_STACK_SIZE);
+
+struct k_mutex mutex;
+struct k_sem sem;
+struct k_timer timer;
+
+void clean_up(struct k_timer *timer)
```
Comment (enjiamai) [inline tests/stress_tests/mutex_stress/src/main.c:21]: for helper function, we better make them static.

File: tests/stress_tests/mutex_stress/src/main.c [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	int i;
+
+	for (i = 0; i < THREAD_COUNT; i++) {
+		k_thread_abort(&thread_array[i]);
+	}
+	k_thread_abort(&sem_thread);
+
+}
+
+void thread_func(void *p1, void *p2, void *p3)
```
Comment (enjiamai) [inline tests/stress_tests/mutex_stress/src/main.c:34]: put static also.

File: tests/stress_tests/mutex_stress/src/main.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+				"ret = %d\n", index, ret);
+			ztest_test_fail();
+			clean_up(NULL);
+		}
+
+		k_mutex_unlock(&mutex);
+	}
+}
+
+void sem_give_func(void *p1, void *p2, void *p3)
```
Comment (enjiamai) [inline tests/stress_tests/mutex_stress/src/main.c:62]: static also.

File: tests/stress_tests/mutex_stress/src/main.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+{
+	while (1) {
+		if (sem.count == 0) {
+			k_sem_give(&sem);
+		}
+		k_sleep(K_MSEC(2));
+	}
+}
+
+void test_mutex_stress(void)
```
Comment (enjiamai) [inline tests/stress_tests/mutex_stress/src/main.c:72]: Could you please also put a @brief and @details here to describe what is this test case for? Thanks!

File: tests/stress_tests/mutex_stress/src/main.c [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+struct k_thread thread_array[THREAD_COUNT];
+K_THREAD_STACK_ARRAY_DEFINE(thread_stack, THREAD_COUNT, THREAD_STACK_SIZE);
+
+struct k_thread sem_thread;
+K_THREAD_STACK_DEFINE(sem_thread_stack, THREAD_STACK_SIZE);
+
+struct k_mutex mutex;
+struct k_sem sem;
+struct k_timer timer;
```
Comment (enjiamai) [inline tests/stress_tests/mutex_stress/src/main.c:19]: Make all the global variable static as possible, if they are not used in other place. Just a suggestion.

**Review by enjiamai — CHANGES_REQUESTED** [review]

File: scripts/twister [hunk 6]
```diff
@@ -292,6 +292,7 @@ Artificially long but functional example:
              "Default is to not do any tag filtering. Multiple invocations "
              "are treated as a logical 'or' relationship")
     parser.add_argument("-e", "--exclude-tag", action="append",
+                        default = ["stress"],
```
Comment (nashif) [inline scripts/twister:295]: no please, do not do it like this. mark the test as slow for now

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/stress_tests/mutex_stress/src/main.c [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	}
+}
+
+void sem_give_func(void *p1, void *p2, void *p3)
+{
+	while (1) {
+		if (sem.count == 0) {
+			k_sem_give(&sem);
+		}
+		k_sleep(K_MSEC(2));
```
Comment (maksimmasalski) [inline tests/stress_tests/mutex_stress/src/main.c:68]: Hardcoded value "2" should be placed into a variable. Please also add comment why exactly "2" msec to sleep.

File: tests/stress_tests/mutex_stress/src/main.c [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	k_thread_abort(&sem_thread);
+
+}
+
+void thread_func(void *p1, void *p2, void *p3)
+{
+	int ret, index;
+
+	index = (intptr_t)p1;
+	k_sleep(K_MSEC(100));
```
Comment (maksimmasalski) [inline tests/stress_tests/mutex_stress/src/main.c:39]: Same here, I think better to put that "100" into a variable. Also please add comment why sleep is necessary here.

File: tests/stress_tests/mutex_stress/src/main.c (hunk 5 above)
Comment (chen-png) [reply tests/stress_tests/mutex_stress/src/main.c]: fixed.

File: scripts/twister (hunk 6 above)
Comment (chen-png) [reply scripts/twister]: I have updated to use slow, thanks.

File: tests/stress_tests/mutex_stress/src/main.c (hunk 8 above)
Comment (chen-png) [reply tests/stress_tests/mutex_stress/src/main.c:41]: I added some comments for those sleep operations, thanks.

**Review by enjiamai — APPROVED** [review]

**Pull request closed without merging**